core/src/main/java/org/apache/accumulo/core/fate/zookeeper/ZooUtil.java

```diff
@@ -169,27 +212,198 @@ private static String getFmtTime(final long epoch) {
+        for (int i = 0; i < connectTimeWait / TIME_BETWEEN_CONNECT_CHECKS_MS && tryAgain; i++) {
+          if (zk.getState().isConnected()) {
+            if (instanceSecret != null) {
+              ZooUtil.digestAuth(zk, instanceSecret);
+            }
+            tryAgain = false;
+          } else {
+            UtilWaitThread.sleep(TIME_BETWEEN_CONNECT_CHECKS_MS);
+          }
+        }
+
+      } catch (IOException e) {
+        if (e instanceof UnknownHostException) {
+          /*
+           * Make sure we wait at least as long as the JVM TTL for negative DNS responses
+           */
+          int ttl = AddressUtil.getAddressCacheNegativeTtl((UnknownHostException) e);
+          sleepTime = Math.max(sleepTime, (ttl + 1) * 1000L);
+        }
+        log.warn("Connection to zooKeeper failed, will try again in "
+            + String.format("%.2f secs", sleepTime / 1000.0), e);
+      } finally {
+        if (tryAgain && zk != null) {
+          try {
+            zk.close();
+            zk = null;
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            log.warn("interrupted", e);
+          }
+        }
+      }
+
+      long stopTime = System.nanoTime();
+      long duration = NANOSECONDS.toMillis(stopTime - startTime);
+
+      if (duration > 2L * timeout) {
+        throw new IllegalStateException("Failed to connect to zookeeper (" + connectString
+            + ") within 2x zookeeper timeout period " + timeout);
+      }
+
+      if (tryAgain) {
+        if (2L * timeout < duration + sleepTime + connectTimeWait) {
+          sleepTime = 2L * timeout - duration - connectTimeWait;
+        }
+        if (sleepTime < 0) {
+          connectTimeWait -= sleepTime;
+          sleepTime = 0;
+        }
+        UtilWaitThread.sleep(sleepTime);
+        if (sleepTime < 10000) {
+          sleepTime = sleepTime + (long) (sleepTime * RANDOM.get().nextDouble());
+        }
+      }
     }
+
+    return zk;
   }
 
-  public static void digestAuth(ZooKeeper zoo, String secret) {
-    auth(zoo, "digest", ("accumulo:" + secret).getBytes(UTF_8));
+  /**
+   * Given a zooCache and instanceId, look up the instance name.
+   */
+  public static String getInstanceName(String zooKeepers, int zkSessionTimeout,
+      InstanceId instanceId) {
+    requireNonNull(zooKeepers);
+    var instanceIdBytes = requireNonNull(instanceId).canonical().getBytes(UTF_8);
+    try (var zk = connect("ZooUtil.getInstanceName", zooKeepers, zkSessionTimeout, null)) {
+      for (String name : zk.getChildren(Constants.ZROOT + Constants.ZINSTANCES, false)) {
+        var bytes = zk.getData(Constants.ZROOT + Constants.ZINSTANCES + "/" + name, false, null);
+        if (Arrays.equals(bytes, instanceIdBytes)) {
+          return name;
+        }
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```

Comment:
I think we should not have started doing this because I don't know that we check this and or clear it in a lot of cases. If the condition is not cleared by the code that calls this method, then the condition persists on the thread until it is checked and cleared. It's possible that some library we use may check this condition and do something unintended. I'm wondering if we should just throw InterruptedException.

Reply:
The ZooUtil methods throwing an InterruptedException makes it very messy. While it might be suitable in some places, I think this behavior is better for the Zooutil methods, so we're not passing InterruptedException through hundreds of layers and being stuck not knowing how to handle it. In any case, I don't think this is something I changed... just made it more consistent in the ZooUtil code.

Reply:
> The ZooUtil methods throwing an InterruptedException makes it very messy. While it might be suitable in some places, I think this behavior is better for the Zooutil methods, so we're not passing InterruptedException through hundreds of layers and being stuck not knowing how to handle it.

  Right, you are throwing an IllegalStateException so that we don't have to handle the InterruptedException through the layers. I don't think that calling `Thread.currentThread.interrupt` is required here.

> In any case, I don't think this is something I changed... just made it more consistent in the ZooUtil code.

InstanceOperationsImpl.lookupInstanceName did not call `Thread.currentThread.interrupt()`.

Reply:
> InstanceOperationsImpl.lookupInstanceName did not call `Thread.currentThread.interrupt()`.

By convention, the status is cleared when InterruptedException is thrown/caught. It's widely documented as a best practice to restore it if you're not rethrowing InterruptedException, so that the calling code is capable of knowing that the thread was interrupted, in case it cares. If something cares, then it can handle it. If nothing cares, then it's fine that it persists. The flag is purely informational, and we have no reason to hide the information that the thread had been interrupted from the calling code.

There are several related utility methods that operate in the same scope. One is to lookup the instanceName from the ID, and one is to look up the ID from the instanceName, and another gets the full map. Two of them had this exception handling, which follows widely documented best practices for handling these. Several of these utilities also had redundant implementations, each slightly different. I had to converge on something, so I made the third one consistent with the other two, rather than remove it from them, because those two followed best practices, and this one didn't, and I could think of no good reason to hide the interrupted status from the calling code.

Reply:
Discussed with @dlmarion and have come to the agreement that the current pattern is fine.